File: 22_专题/离线查询/根号分治/SqrtArray/a/SqrtArraySimple.go
package main

import (
	"fmt"
	"math"
	"unsafe"
)

func main() {

}

type SqrtArray[T any] struct {
	n      int
	blocks [][]T
}

// NewSqrtArray 创建一个新的SqrtArray
func NewSqrtArray[T any](n int, f func(int) T, blockSize int) *SqrtArray[T] {
	if n <= 0 {
		return &SqrtArray[T]{}
	}

	blockCount := int(math.Sqrt(float64(n)))
	if blockSize < 1 {
		blockSize = (n + blockCount - 1) / blockCount
	}

	blocks := make([][]T, blockCount)
	for i := 0; i < blockCount; i++ {
		start, end := i*blockSize, min((i+1)*blockSize, n)
		blocks[i] = make([]T, 0, end-start)
		for j := start; j < end; j++ {
			blocks[i] = append(blocks[i], f(j))
		}
	}

	return &SqrtArray[T]{
		n:      n,
		blocks: blocks,
	}
}

func (s *SqrtArray[T]) Set(i int, v T) {
	if i < 0 || i >= s.n {
		panic("index out of range")
	}
	bid, pos := s.findKth(i)
	s.blocks[bid][pos] = v
}

func (s *SqrtArray[T]) Get(i int) T {
	if i < 0 {
		i += s.n
	}
	if i < 0 || i >= s.n {
		panic("index out of range")
	}
	bid, pos := s.findKth(i)
	return s.blocks[bid][pos]
}

// At 获取指定索引的值，支持负数索引
func (s *SqrtArray[T]) At(i int) (T, bool) {
	if i < 0 {
		i += s.n
	}
	if i < 0 || i >= s.n {
		var zero T
		return zero, false
	}
	return s.Get(i), true
}

// Push 在数组末尾添加元素
func (s *SqrtArray[T]) Push(v T) {
	s.Insert(s.n, v)
}

// Pop 删除并返回指定索引的元素
func (s *SqrtArray[T]) Pop(i int) (T, bool) {
	if i < 0 {
		i += s.n
	}
	if i < 0 || i >= s.n {
		var zero T
		return zero, false
	}

	var bi int
	var res T
	if i == s.n-1 {
		bi = len(s.blocks) - 1
		res = s.blocks[bi][len(s.blocks[bi])-1]
		s.blocks[bi] = s.blocks[bi][:len(s.blocks[bi])-1]
	} else {
		for ; i >= len(s.blocks[bi]); i -= len(s.blocks[bi]) {
			bi++
		}
		res = s.blocks[bi][i]
		s.blocks[bi] = append(s.blocks[bi][:i], s.blocks[bi][i+1:]...)
	}
	s.n--
	if len(s.blocks[bi]) == 0 {
		s.blocks = append(s.blocks[:bi], s.blocks[bi+1:]...)
	}
	return res, true
}

// Shift 删除并返回数组第一个元素
func (s *SqrtArray[T]) Shift() (T, bool) {
	return s.Pop(0)
}

// Unshift 在数组开头添加元素
func (s *SqrtArray[T]) Unshift(v T) {
	s.Insert(0, v)
}

// Erase 删除区间 [start, end) 内的元素
func (s *SqrtArray[T]) Erase(start, end int) {
	if start < 0 {
		start = 0
	}
	if end > s.n {
		end = s.n
	}
	if start >= end {
		return
	}

	bid, startPos := s.findKth(start)
	deleteCount := end - start
	for bid < len(s.blocks) && deleteCount > 0 {
		block := s.blocks[bid]
		endPos := min(len(block), startPos+deleteCount)
		curDeleteCount := endPos - startPos
		if curDeleteCount == len(block) {
			s.blocks = append(s.blocks[:bid], s.blocks[bid+1:]...)
			bid--
		} else {
			s.blocks[bid] = append(block[:startPos], block[endPos:]...)
		}
		deleteCount -= curDeleteCount
		s.n -= curDeleteCount
		startPos = 0
		bid++
	}
}

// Insert 在指定位置插入元素
func (s *SqrtArray[T]) Insert(i int, v T) {
	if s.n == 0 {
		s.blocks = append(s.blocks, []T{v})
		s.n++
		return
	}

	bi := 0
	if i >= s.n {
		bi = len(s.blocks) - 1
		s.blocks[bi] = append(s.blocks[bi], v)
	} else {
		for ; bi < len(s.blocks) && i >= len(s.blocks[bi]); i -= len(s.blocks[bi]) {
			bi++
		}
		s.blocks[bi] = append(s.blocks[bi][:i], append([]T{v}, s.blocks[bi][i:]...)...)
	}
	s.n++

	sqrtn2 := int(math.Sqrt(float64(s.n))) * 3
	// 定期重构
	if len(s.blocks[bi]) > 2*sqrtn2 {
		y := s.blocks[bi][sqrtn2:]
		s.blocks[bi] = s.blocks[bi][:sqrtn2]
		s.blocks = append(s.blocks[:bi+1], append([][]T{y}, s.blocks[bi+1:]...)...)
	}
}

// Enumerate 遍历区间 [start, end) 内的元素，并选择是否在遍历后删除
func (s *SqrtArray[T]) Enumerate(start, end int, f func(T), erase bool) {
	bid, startPos := s.findKth(start)
	count := end - start

	for bid < len(s.blocks) && count > 0 {
		block := s.blocks[bid]
		endPos := min(len(block), startPos+count)
		for j := startPos; j < endPos; j++ {
			f(block[j])
		}

		curDeleteCount := endPos - startPos
		if erase {
			if curDeleteCount == len(block) {
				s.blocks = append(s.blocks[:bid], s.blocks[bid+1:]...)
				bid--
			} else {
				s.blocks[bid] = append(block[:startPos], block[endPos:]...)
			}
			s.n -= curDeleteCount
		}

		count -= curDeleteCount
		startPos = 0
		bid++
	}
}

// Slice 返回指定区间的切片
func (s *SqrtArray[T]) Slice(start, end int) []T {
	if start < 0 {
		start = 0
	}
	if end > s.n {
		end = s.n
	}
	if start >= end {
		return []T{}
	}
	count := end - start
	res := make([]T, count)
	bid, startPos := s.findKth(start)
	ptr := 0
	for bid < len(s.blocks) && count > 0 {
		block := s.blocks[bid]
		endPos := min(len(block), startPos+count)
		curCount := endPos - startPos
		copy(res[ptr:], block[startPos:endPos])
		ptr += curCount
		count -= curCount
		startPos = 0
		bid++
	}
	return res
}

// Fill 用指定值填充整个数组
func (s *SqrtArray[T]) Fill(v T) {
	for i := range s.blocks {
		for j := range s.blocks[i] {
			s.blocks[i][j] = v
		}
	}
}

// ISlice 返回一个迭代器，用于遍历区间 [start, end) 内的元素
func (s *SqrtArray[T]) ISlice(start, end int, reverse bool) <-chan T {
	ch := make(chan T)
	go func() {
		defer close(ch)
		if start < 0 {
			start = 0
		}
		if end > s.n {
			end = s.n
		}
		if start >= end {
			return
		}
		count := end - start

		if reverse {
			bid, endPos := s.findKth(end - 1)
			for bid >= 0 && count > 0 {
				block := s.blocks[bid]
				startPos := max(0, endPos-count)
				curCount := endPos - startPos
				for j := endPos - 1; j >= startPos; j-- {
					ch <- block[j]
				}
				count -= curCount
				bid--
				if bid >= 0 {
					endPos = len(s.blocks[bid])
				}
			}
		} else {
			bid, startPos := s.findKth(start)
			for bid < len(s.blocks) && count > 0 {
				block := s.blocks[bid]
				endPos := min(len(block), startPos+count)
				curCount := endPos - startPos
				for j := startPos; j < endPos; j++ {
					ch <- block[j]
				}
				count -= curCount
				startPos = 0
				bid++
			}
		}
	}()
	return ch
}

// ForEach 遍历整个数组
func (s *SqrtArray[T]) ForEach(callback func(T, int)) {
	ptr := 0
	for bi := 0; bi < len(s.blocks); bi++ {
		for j := 0; j < len(s.blocks[bi]); j++ {
			callback(s.blocks[bi][j], ptr)
			ptr++
		}
	}
}

// Entries 返回一个迭代器，用于遍历数组的索引和值
func (s *SqrtArray[T]) Entries() <-chan [2]interface{} {
	ch := make(chan [2]interface{})
	go func() {
		defer close(ch)
		ptr := 0
		for i := 0; i < len(s.blocks); i++ {
			block := s.blocks[i]
			for j := 0; j < len(block); j++ {
				ch <- [2]interface{}{ptr, block[j]}
				ptr++
			}
		}
	}()
	return ch
}

// Iterator 返回一个迭代器，用于遍历数组的值
func (s *SqrtArray[T]) Iterator() <-chan T {
	ch := make(chan T)
	go func() {
		defer close(ch)
		for i := 0; i < len(s.blocks); i++ {
			block := s.blocks[i]
			for j := 0; j < len(block); j++ {
				ch <- block[j]
			}
		}
	}()
	return ch
}

func (s *SqrtArray[T]) String() string {
	return fmt.Sprintf("SqrtArray{%v}", s.blocks)
}

func (s *SqrtArray[T]) findKth(index int) (int, int) {
	var bi, pos int
	if index < s.n>>1 {
		bi, pos = s.findFromStart(index)
	} else {
		bi, pos = s.findFromEnd(s.n - index - 1)
	}
	return bi, pos
}

func (s *SqrtArray[T]) findFromStart(step int) (bi, pos int) {
	for j := 0; j < len(s.blocks); j++ {
		if step < len(s.blocks[j]) {
			return j, step
		}
		step -= len(s.blocks[j])
	}
	panic("index out of range")
}

func (s *SqrtArray[T]) findFromEnd(step int) (bi, pos int) {
	for j := len(s.blocks) - 1; j >= 0; j-- {
		if step < len(s.blocks[j]) {
			return j, len(s.blocks[j]) - 1 - step
		}
		step -= len(s.blocks[j])
	}
	panic("index out of range")
}

func (s *SqrtArray[T]) Len() int {
	return s.n
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Replace replaces the elements s[i:j] by the given v, and returns the modified slice.
// !Like JavaScirpt's Array.prototype.splice.
func Replace[S ~[]E, E any](s S, i, j int, v ...E) S {
	if i < 0 {
		i = 0
	}
	if j > len(s) {
		j = len(s)
	}
	if i == j {
		return Insert(s, i, v...)
	}
	if j == len(s) {
		return append(s[:i], v...)
	}
	tot := len(s[:i]) + len(v) + len(s[j:])
	if tot > cap(s) {
		s2 := append(s[:i], make(S, tot-i)...)
		copy(s2[i:], v)
		copy(s2[i+len(v):], s[j:])
		return s2
	}
	r := s[:tot]
	if i+len(v) <= j {
		copy(r[i:], v)
		copy(r[i+len(v):], s[j:])
		// clear(s[tot:])
		return r
	}
	if !overlaps(r[i+len(v):], v) {
		copy(r[i+len(v):], s[j:])
		copy(r[i:], v)
		return r
	}
	y := len(v) - (j - i)
	if !overlaps(r[i:j], v) {
		copy(r[i:j], v[y:])
		copy(r[len(s):], v[:y])
		rotateRight(r[i:], y)
		return r
	}
	if !overlaps(r[len(s):], v) {
		copy(r[len(s):], v[:y])
		copy(r[i:j], v[y:])
		rotateRight(r[i:], y)
		return r
	}
	k := startIdx(v, s[j:])
	copy(r[i:], v)
	copy(r[i+len(v):], r[i+k:])
	return r
}

func rotateLeft[E any](s []E, r int) {
	for r != 0 && r != len(s) {
		if r*2 <= len(s) {
			swap(s[:r], s[len(s)-r:])
			s = s[:len(s)-r]
		} else {
			swap(s[:len(s)-r], s[r:])
			s, r = s[len(s)-r:], r*2-len(s)
		}
	}
}

func rotateRight[E any](s []E, r int) {
	rotateLeft(s, len(s)-r)
}

func swap[E any](x, y []E) {
	for i := 0; i < len(x); i++ {
		x[i], y[i] = y[i], x[i]
	}
}

func overlaps[E any](a, b []E) bool {
	if len(a) == 0 || len(b) == 0 {
		return false
	}
	elemSize := unsafe.Sizeof(a[0])
	if elemSize == 0 {
		return false
	}
	return uintptr(unsafe.Pointer(&a[0])) <= uintptr(unsafe.Pointer(&b[len(b)-1]))+(elemSize-1) &&
		uintptr(unsafe.Pointer(&b[0])) <= uintptr(unsafe.Pointer(&a[len(a)-1]))+(elemSize-1)
}

func startIdx[E any](haystack, needle []E) int {
	p := &needle[0]
	for i := range haystack {
		if p == &haystack[i] {
			return i
		}
	}
	panic("needle not found")
}

func Insert[S ~[]E, E any](s S, i int, v ...E) S {
	if i < 0 {
		i = 0
	}
	if i > len(s) {
		i = len(s)
	}

	m := len(v)
	if m == 0 {
		return s
	}
	n := len(s)
	if i == n {
		return append(s, v...)
	}
	if n+m > cap(s) {
		s2 := append(s[:i], make(S, n+m-i)...)
		copy(s2[i:], v)
		copy(s2[i+m:], s[i:])
		return s2
	}
	s = s[:n+m]
	if !overlaps(v, s[i+m:]) {
		copy(s[i+m:], s[i:])
		copy(s[i:], v)
		return s
	}
	copy(s[n:], v)
	rotateRight(s[i:], m)
	return s
}

File: 22_专题/离线查询/根号分治/SqrtArray/SqrtArraySimple.ts
/* eslint-disable generator-star-spacing */
/* eslint-disable no-empty */

// https://github.com/spaghetti-source/algorithm/blob/4fdac8202e26def25c1baf9127aaaed6a2c9f7c7/data_structure/sqrt_array.cc
//
// SQRT Array
//
// Description:
//   An array with O(sqrt(n)) deletion and insertion
//
// Algorithm:
//   Decompose array into O(sqrt(n)) subarrays.
//   Then the all operation is performed in O(sqrt(n)).
//
// Complexity:
//   O(sqrt(n)); however, due to the cheap constant factor,
//   it is comparable with binary search trees.
//   If only deletion is required, it is better choice.
//
// 优化点：
// 1. `findKth`使用二分查找或更高效的方式查找块，以替代线性查找(见golang版本);
// 2. 删除后，合并相邻的小块，以减少块的数量。

/**
 * 分块数组.
 */
class SqrtArray<T = number> {
  private _n = 0
  private readonly _blocks: T[][] = []
  private readonly _blockSize: number

  constructor(n: number, f: (i: number) => T, blockSize = 1 << 9) {
    const blockCount = ((n + blockSize - 1) / blockSize) | 0
    const blocks: T[][] = Array(blockCount)
    for (let i = 0; i < blockCount; i++) {
      const start = i * blockSize
      const end = Math.min((i + 1) * blockSize, n)
      const cur = Array(end - start)
      for (let j = start; j < end; j++) cur[j - start] = f(j)
      blocks[i] = cur
    }

    this._n = n
    this._blocks = blocks
    this._blockSize = blockSize
  }

  /**
   * 0<= i < {@link length}
   */
  set(i: number, v: T): void {
    const { bid, pos } = this._findKth(i)
    this._blocks[bid][pos] = v
  }

  /**
   * 0<= i < {@link length}
   */
  get(i: number): T | undefined {
    const { bid, pos } = this._findKth(i)
    return this._blocks[bid][pos]
  }

  /**
   * i可以是负数索引.
   */
  at(i: number): T | undefined {
    if (i < 0) i += this._n
    if (i < 0 || i >= this._n) return undefined
    return this.get(i)
  }

  push(v: T): void {
    this.insert(this._n, v)
  }

  /**
   * i可以是负数索引.
   */
  pop(i = this._n - 1): T | undefined {
    if (i < 0) i += this._n
    if (i < 0 || i >= this._n) return undefined
    const { bid, pos } = this._findKth(i)
    const res = this._blocks[bid].splice(pos, 1)[0]
    this._n--
    // TODO: 过稀疏时，合并相邻块
    if (!this._blocks[bid].length) this._blocks.splice(bid, 1)
    return res
  }

  shift(): T | undefined {
    return this.pop(0)
  }

  unshift(v: T): void {
    this.insert(0, v)
  }

  /**
   * 删除区间 [start, end) 内的元素.
   * 0<= start <= end <= {@link length}
   */
  erase(start: number, end: number): void {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return

    let { bid, pos } = this._findKth(start)
    let deleteCount = end - start
    let eraseStart = -1
    let eraseCount = 0
    for (; bid < this._blocks.length && deleteCount > 0; bid++) {
      const block = this._blocks[bid]
      const endPos = Math.min(block.length, pos + deleteCount)
      const curDeleteCount = endPos - pos
      if (curDeleteCount === block.length) {
        if (eraseStart === -1) eraseStart = bid
        eraseCount++
      } else {
        block.splice(pos, curDeleteCount)
      }
      deleteCount -= curDeleteCount
      this._n -= curDeleteCount
      pos = 0
    }
    if (eraseStart !== -1) this._blocks.splice(eraseStart, eraseCount)
  }

  /**
   * 在 i 位置`前`插入 v.
   * 0<= i <= {@link length}
   */
  insert(i: number, v: T): void {
    if (!this._n) {
      this._blocks.push([v])
      this._n++
      return
    }

    let bid = 0
    if (i >= this._n) {
      bid = this._blocks.length - 1
      this._blocks[bid].push(v)
    } else {
      const { bid: bid_, pos } = this._findKth(i)
      bid = bid_
      this._blocks[bid_].splice(pos, 0, v)
    }
    this._n++

    // 定期重构
    if (this._blocks[bid].length >>> 1 > this._blockSize) {
      const y = this._blocks[bid].splice(this._blockSize)
      this._blocks.splice(bid + 1, 0, y)
    }
  }

  /**
   * 遍历区间 [start, end) 内的元素,并选择是否在遍历后删除.
   * 0<= start <= end <= {@link length}
   */
  enumerate(start: number, end: number, f: (value: T) => void, erase = false): void {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return

    let { bid, pos } = this._findKth(start)
    let count = end - start

    let eraseStart = -1
    let eraseCount = 0
    for (; bid < this._blocks.length && count > 0; bid++) {
      const block = this._blocks[bid]
      const endPos = Math.min(block.length, pos + count)
      for (let j = pos; j < endPos; j++) {
        f(block[j])
      }

      const curDeleteCount = endPos - pos
      if (erase) {
        if (curDeleteCount === block.length) {
          if (eraseStart === -1) eraseStart = bid
          eraseCount++
        } else {
          block.splice(pos, curDeleteCount)
        }
        this._n -= curDeleteCount
      }

      count -= curDeleteCount
      pos = 0
    }

    if (erase && eraseStart !== -1) this._blocks.splice(eraseStart, eraseCount)
  }

  slice(start: number, end: number): T[] {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return []
    let count = end - start
    const res: T[] = Array(count)
    let { bid, pos } = this._findKth(start)
    let ptr = 0
    for (; bid < this._blocks.length && count > 0; bid++) {
      const block = this._blocks[bid]
      const endPos = Math.min(block.length, pos + count)
      const curCount = endPos - pos
      for (let j = pos; j < endPos; j++) res[ptr++] = block[j]
      count -= curCount
      pos = 0
    }
    return res
  }

  fill(v: T): this {
    this._blocks.forEach(b => b.fill(v))
    return this
  }

  /**
   * 返回一个迭代器，用于遍历区间 [start, end) 内的元素.
   */
  *islice(start: number, end: number, reverse = false): IterableIterator<T> {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return
    let count = end - start

    if (reverse) {
      let { bid, pos } = this._findKth(end - 1)
      for (; ~bid && count > 0; bid--, ~bid && (pos = this._blocks[bid].length)) {
        const block = this._blocks[bid]
        const startPos = Math.max(0, pos - count)
        const curCount = pos - startPos
        for (let j = pos - 1; j >= startPos; j--) {
          yield block[j]
        }
        count -= curCount
      }
    } else {
      let { bid, pos } = this._findKth(start)
      for (; bid < this._blocks.length && count > 0; bid++) {
        const block = this._blocks[bid]
        const endPos = Math.min(block.length, pos + count)
        const curCount = endPos - pos
        for (let j = pos; j < endPos; j++) {
          yield block[j]
        }
        count -= curCount
        pos = 0
      }
    }
  }

  forEach(callback: (value: T, index: number) => void): void {
    let ptr = 0
    for (let bi = 0; bi < this._blocks.length; ++bi) {
      for (let j = 0; j < this._blocks[bi].length; ++j) {
        callback(this._blocks[bi][j], ptr++)
      }
    }
  }

  *entries(): IterableIterator<[number, T]> {
    let ptr = 0
    for (let i = 0; i < this._blocks.length; i++) {
      const block = this._blocks[i]
      for (let j = 0; j < block.length; j++) {
        yield [ptr++, block[j]]
      }
    }
  }

  *[Symbol.iterator](): Iterator<T> {
    for (let i = 0; i < this._blocks.length; i++) {
      const block = this._blocks[i]
      for (let j = 0; j < block.length; j++) {
        yield block[j]
      }
    }
  }

  toString(): string {
    return `SqrtArray{${this._blocks}}`
  }

  private _findKth(index: number): { bid: number; pos: number } {
    if (index < this._n >>> 1) return this._findFromStart(index)
    return this._findFromEnd(this._n - index - 1)
  }

  private _findFromStart(step: number): { bid: number; pos: number } {
    for (let i = 0; i < this._blocks.length; i++) {
      const block = this._blocks[i]
      if (step < block.length) {
        return { bid: i, pos: step }
      }
      step -= block.length
    }
    throw new Error('index out of range')
  }

  private _findFromEnd(step: number): { bid: number; pos: number } {
    for (let i = this._blocks.length - 1; ~i; i--) {
      const block = this._blocks[i]
      if (step < block.length) {
        return { bid: i, pos: block.length - step - 1 }
      }
      step -= block.length
    }
    throw new Error('index out of range')
  }

  get length(): number {
    return this._n
  }
}

export { SqrtArray }

if (require.main === module) {
  const arr = new SqrtArray<number>(0, i => i, 1 << 9)
  const n = 1e6
  const rands = Array(n)
    .fill(0)
    .map((_, i) => Math.floor(Math.random() * 1e9))
  console.time('insert')
  for (let i = 0; i < n; i++) {
    arr.insert(n - i, i)
    arr.get(i)
    arr.set(i, i)
    arr.unshift(i)
    arr.shift()
  }

  for (let i = 0; i < n; i++) {
    arr.get(i)
  }
  for (let i = 0; i < n; i++) {
    arr.pop(0)
  }

  console.timeEnd('insert')
  console.log(rands.slice(0, 10))
  arr.erase(0, 100)
  console.log(arr.length)

  // https://leetcode.cn/problems/design-circular-deque/
  class MyCircularDeque {
    private readonly _sqrt: SqrtArray<number>
    private readonly _k: number
    constructor(k: number) {
      this._sqrt = new SqrtArray(0, () => 0)
      this._k = k
    }

    insertFront(value: number): boolean {
      if (this.isFull()) return false
      this._sqrt.unshift(value)
      return true
    }

    insertLast(value: number): boolean {
      if (this.isFull()) return false
      this._sqrt.push(value)
      return true
    }

    deleteFront(): boolean {
      if (this.isEmpty()) return false
      this._sqrt.shift()
      return true
    }

    deleteLast(): boolean {
      if (this.isEmpty()) return false
      this._sqrt.pop()
      return true
    }

    getFront(): number {
      return this.isEmpty() ? -1 : this._sqrt.get(0)!
    }

    getRear(): number {
      return this.isEmpty() ? -1 : this._sqrt.get(this._sqrt.length - 1)!
    }

    isEmpty(): boolean {
      return this._sqrt.length === 0
    }

    isFull(): boolean {
      return this._sqrt.length === this._k
    }
  }
}
